Round BigDecimals with too negative exp

leoyvens opened the pull request:
Numbers with too much precision after the dot can be rounded to an acceptable amount of precision. This should fix subgraphs with numbers which would otherwise not pass the validation due to a too negative exponent.

That3Percent commented on graph/src/data/store/scalar.rs:
```diff
@@ -19,7 +19,15 @@ use std::str::FromStr;
-/// All operations on `BigDecimal` return a normalized value.
+/// These are the limits of IEEE-754 decimal128, a format we may want to switch to. See
+/// https://en.wikipedia.org/wiki/Decimal128_floating-point_format.
+pub const BIG_DECIMAL_MIN_EXP: i32 = -6143;
```
This is nit-picky, but these can be made associated constants and not pollute the already large prelude namespace:
```rust
impl BigDecimal {
    pub const MIN_EXP: i32 = -6143;
}

fn test() {
   let exp = BigDecimal::MIN_EXP;
}
```

leoyvens replied:
That really is nicer!

That3Percent commented on tests/integration-tests/big-decimal/test/test.js:
```diff
@@ -0,0 +1,77 @@
+const path = require("path");
```
I really like that we moved to integration tests, and that I consistently see new integration tests being added with your changes.

The amount of code that these require as compared to the size of the change is intimidating. Is there a way to move a lot of this into a library or something?

leoyvens replied:
I carn try to move the contract, migration, and other files that repeat between many tests to a shared directory. The `waitForSubgraphToBeSynced` function can also be shared.

leoyvens commented:
@That3Percent Review addressed, I've moved contracts and migrations which are reused under `integration-tests/common`, which is a start to reduce duplication in tests.

leoyvens commented:
I've changed this to always round to 34 decimal digits (the test was also updated). This change may affect users, but it's hard to evaluate whether this rounding will have undesirable effects to a subgraph without knowing the requirements of the applications, so I suggest that we try it out and see if users report issues.

I also added normalization after the BigDecimal math operations.

That3Percent commented on graph/src/data/store/scalar.rs:
```diff
@@ -71,20 +69,18 @@ impl BigDecimal {
-        let (bigint, exp) = self.0.as_bigint_and_exponent();
+
+        // Round to the maximum significant digits.
+        let big_decimal = self.0.with_prec(Self::MAX_SIGNFICANT_DIGITS as u64);
```
When switching to decimal128 we may need to remember that this method has a specific rounding mode and we need to match it or make a version bump.

leoyvens replied:
Hopefully there isn't too much variation in how to round a decimal number, but versioning might be the best anyways to make sure we don't miss any subtle behavior difference.

That3Percent commented on tests/integration-tests/big-decimal/src/mapping.ts:
```diff
@@ -6,21 +6,27 @@ export function handleTrigger(event: Trigger): void {
+  let big = BigDecimal.fromString("99999999999999999999999999999999999");
+
+  // This has 35 zeros.
+  let rounded = BigDecimal.fromString("100000000000000000000000000000000000");
```
This test case doesn't prove as much as it could. It would be better to use `8` instead of `9` to demonstrate exactly where the rounding occurs.

leoyvens replied:
I've added that test case.

leoyvens replied:
I've added that test case.

That3Percent commented on graph/src/data/store/scalar.rs:
```diff
@@ -115,23 +126,23 @@ impl Add for BigDecimal {
 
     fn sub(self, other: Self) -> Self {
-        Self::from(self.0.sub(other.0))
+        Self::from(self.0.sub(other.0)).normalized()
```
`from` already normalizes - so all these math operations now normalize twice.

leoyvens replied:
True, I got confused here and thought we weren't normalizing the output. Removed this commit.
